File: web_app/motion_detector/src/models/MotionEvent.js
import { format } from 'date-fns';

/**
 * Model class representing a motion detection event
 */
class MotionEvent {
  /**
   * Creates a new motion event
   * @param {string} id - Unique identifier for the event
   * @param {Date} timestamp - Timestamp when the motion was detected
   * @param {string} deviceId - ID of the device that detected the motion
   * @param {number} temperature - Temperature reading at the time of the event (in Celsius)
   * @param {number} humidity - Humidity reading at the time of the event (in percentage)
   * @param {string} imageUrl - URL to the image captured during the motion event
   */
  constructor(id, timestamp, deviceId, temperature, humidity, imageUrl) {
    this.id = id;
    this.timestamp = timestamp;
    this.deviceId = deviceId;
    this.temperature = temperature;
    this.humidity = humidity;
    this.imageUrl = imageUrl;
  }

  /**
   * Creates a formatted time string from the timestamp with AM/PM
   * @returns {string} Formatted time string
   */
  get formattedTime() {
    return format(this.timestamp, 'h:mm a');
  }

  /**
   * Creates a formatted date string from the timestamp
   * @returns {string} Formatted date string
   */
  get formattedDate() {
    return format(this.timestamp, 'dd/MM/yyyy');
  }
}

export default MotionEvent;

File: lib/models/motion_event.dart
/// Model class representing a motion detection event
class MotionEvent {
  /// Unique identifier for the event
  final String id;

  /// Timestamp when the motion was detected
  final DateTime timestamp;

  /// ID of the device that detected the motion
  final String deviceId;

  /// Temperature reading at the time of the event (in Celsius)
  final double temperature;

  /// Humidity reading at the time of the event (in percentage)
  final double humidity;

  /// URL to the image captured during the motion event
  final String imageUrl;

  /// Creates a new motion event
  const MotionEvent({
    required this.id,
    required this.timestamp,
    required this.deviceId,
    required this.temperature,
    required this.humidity,
    required this.imageUrl,
  });

  /// Creates a formatted time string from the timestamp with AM/PM
  String get formattedTime {
    final hour = timestamp.hour > 12 ? timestamp.hour - 12 : timestamp.hour;
    final hourDisplay = hour == 0 ? 12 : hour; // Handle midnight (0:00) as 12 AM
    final amPm = timestamp.hour >= 12 ? 'PM' : 'AM';
    return '${hourDisplay.toString().padLeft(2, '0')}:${timestamp.minute.toString().padLeft(2, '0')} $amPm';
  }

  /// Creates a formatted date string from the timestamp
  String get formattedDate {
    return '${timestamp.day.toString().padLeft(2, '0')}/${timestamp.month.toString().padLeft(2, '0')}/${timestamp.year}';
  }
}
